pkg/storage/utils/metadata/cs3.go

```diff
@@ -258,49 +286,156 @@ func (cs3 *CS3) SimpleDownload(ctx context.Context, downloadpath string) (conten
 	}
-
-	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
 	if err != nil {
-		return []byte{}, err
+		log.Error().Err(err).Msg("SimpleDownload failed. Giving up.")
+		return []byte{}, nil
 	}
 
-	md, _ := metadata.FromOutgoingContext(ctx)
-	req.Header.Add(ctxpkg.TokenHeader, md.Get(ctxpkg.TokenHeader)[0])
-	resp, err := cs3.dataGatewayClient.Do(req)
+	return b, nil
+}
+
+// Download reads a file from the metadata storage
+func (cs3 *CS3) Download(ctx context.Context, req DownloadRequest) (*DownloadResponse, error) {
+	ctx, span := tracer.Start(ctx, "Download")
+	defer span.End()
+
+	client, err := cs3.providerClient()
 	if err != nil {
-		return []byte{}, err
+		return nil, err
 	}
-
-	if resp.StatusCode != http.StatusOK {
-		return []byte{}, errtypes.NotFound(dreq.Ref.Path)
+	ctx, err = cs3.getAuthContext(ctx)
+	if err != nil {
+		return nil, err
 	}
 
-	b, err := io.ReadAll(resp.Body)
-	if err != nil {
-		return []byte{}, err
+	dreq := provider.InitiateFileDownloadRequest{
+		Ref: &provider.Reference{
+			ResourceId: cs3.SpaceRoot,
+			Path:       utils.MakeRelativePath(req.Path),
+		},
 	}
+	// FIXME add a dedicated property on the CS3 InitiateFileDownloadRequest message
+	// well the gateway never forwards the initiate request to the storageprovider, so we have to send it in the actual GET request
+	// if len(req.IfNoneMatch) > 0 {
+	//   dreq.Opaque = utils.AppendPlainToOpaque(dreq.Opaque, "if-none-match", strings.Join(req.IfNoneMatch, ","))
+	// }
 
-	if err = resp.Body.Close(); err != nil {
-		return []byte{}, err
+	dres := DownloadResponse{}
+
+	downloadFunc := func() error {
+		res, err := client.InitiateFileDownload(ctx, &dreq)
+		if err != nil {
+			return errtypes.NotFound(dreq.Ref.Path)
+		}
+
+		var endpoint string
+
+		for _, proto := range res.GetProtocols() {
+			if proto.Protocol == "spaces" {
+				endpoint = proto.GetDownloadEndpoint()
+				break
+			}
+		}
+		if endpoint == "" {
+			return errors.New("metadata storage doesn't support the spaces download protocol")
+		}
+
+		hreq, err := http.NewRequest(http.MethodGet, endpoint, nil)
+		if err != nil {
+			return err
+		}
+
+		for _, etag := range req.IfNoneMatch {
+			hreq.Header.Add(net.HeaderIfNoneMatch, etag)
+		}
+
+		md, _ := metadata.FromOutgoingContext(ctx)
+		hreq.Header.Add(ctxpkg.TokenHeader, md.Get(ctxpkg.TokenHeader)[0])
+		resp, err := cs3.dataGatewayClient.Do(hreq)
+		if err != nil {
+			return err
+		}
+
+		dres.Etag = resp.Header.Get("etag")
+		dres.Etag = resp.Header.Get("oc-etag") // takes precedence
+
+		if err := errtypes.NewErrtypeFromHTTPStatusCode(resp.StatusCode, hreq.URL.Path); err != nil {
+			return err
+		}
+
+		dres.Mtime, err = time.Parse(time.RFC1123Z, resp.Header.Get("last-modified"))
+		if err != nil {
+			return err
+		}
+
+		dres.Content, err = io.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+
+		if err = resp.Body.Close(); err != nil {
+			return err
+		}
+		return nil
+	}
+	err = downloadFunc()
+	// TODO add retries
```

Comment:
not needed here?

Reply:
No, not really ... I put it there because we added `Download()` with more options. And one option in the future could be built in retries ... we can drop the comment.